### PR TITLE
rgw : remove key-marker when list versions

### PR DESCRIPTION
**Pull request by BryceCao**
Fixes: https://tracker.ceph.com/issues/47984
Signed-off-by: caolei <halei15848934852@163.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_rados.cc [hunk 1]
```diff
@@ -1894,6 +1894,14 @@ int RGWRados::Bucket::List::list_objects_ordered(
 	next_marker = index_key;
       }
 
+      if (cur_marker.instance.empty()) {
+        if (cur_marker.name == index_key.name) continue;
```
Comment (ivancich) [inline src/rgw/rgw_rados.cc:1898]: At least when there is no instance in the marker (i.e., instance is empty), the listing should start after the marker. Do you have an example where that's not the case?

In the case where there is an instance in the marker, again I would expect the listing to start *after* the name/instance. Do you have any output where that's not the case?

**Review by ivancich — COMMENTED** [review]
Something does not seem right with respect to how cls_bucket_list_ordered and deeper layers are supposed to work. We may need to resolve this deeper in the call stack.

Please provide more information as to what you are seeing. Thanks!

**Comment by ivancich** [comment]
@BryceCao Your trackers tend not to be filled out correctly in two respects. First, the "Pull request ID" field should contain the PR number; that's how you link the tracker to the PR, not by adding a link as a comment. Second, the "Backport" field needs to be filled out if backports are expected, and the backport team will do the work.

Thanks for all your contributions!

**Comment by ivancich** [comment]
@BryceCao Since this is important to get right, do you think you could add a unit test that would a) reveal the incorrect behavior and b) show that the solution is working correctly?

**Comment by BryceCao** [comment]
@ivancich , my ceph cluster is based on master code, as `ceph version 16.0.0-6824-ga663b3e65f (a663b3e65f1cd0e99f1922bd6de84b2d27900267) pacific (dev)`, and i create a bucket, put a obj. Then open bucket versioning, and put the obj again. Firstly, i list bucket version when specifide the keymarker to the obj name  only, but all versions of the obj were listed. as 
```
GET /test2/?versions&key-marker=a&encoding-type=url HTTP/1.1
Host: 10.221.128.15:8080
x-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
Authorization: AWS4-HMAC-SHA256 Credential=cl/20201113//s3/aws4_request, SignedHeaders=amz-sdk-invocation-id;amz-sdk-retry;content-type;host;user-agent;x-amz-content-sha256;x-amz-date, Signature=7d21e68ee770fd746fd8a693e661cbe3b20216b920df5a838dd4977938cdf0dc
X-Amz-Date: 20201113T100857Z
User-Agent: aws-sdk-java/1.11.418 Windows_10/10.0 OpenJDK_64-Bit_Server_VM/25.212-b04 java/1.8.0_212-3-redhat
amz-sdk-invocation-id: 77d17b23-ebf7-e549-3db9-916a8a970d74
amz-sdk-retry: 0/0/500
Content-Type: application/octet-stream
Content-Length: 0
Connection: Keep-Alive

HTTP/1.1 200 OK
Transfer-Encoding: chunked
x-amz-request-id: tx000000000000000000015-005fae5b2d-5e98-default
Content-Type: application/xml
Date: Fri, 13 Nov 2020 10:08:45 GMT
Connection: Keep-Alive

3d0
<?xml version="1.0" encoding="UTF-8"?><ListVersionsResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><EncodingType>url</EncodingType><Name>test2</Name><Prefix></Prefix><MaxKeys>1000</MaxKeys><IsTruncated>false</IsTruncated><KeyMarker>a</KeyMarker><VersionIdMarker></VersionIdMarker><Version><Key>a</Key><VersionId>eaVaaCTrIxDnI3ku8FhQ1m.CMPAv7gU</VersionId><IsLatest>true</IsLatest><LastModified>2020-11-13T10:08:44.357Z</LastModified><ETag>&quot;c662c33d7b4d560f87a167ccca9f5239&quot;</ETag><Size>439</Size><StorageClass>STANDARD</StorageClass><Owner><ID>cl</ID><DisplayName>cl</DisplayName></Owner><Type>Normal</Type></Version><Version><Key>a</Key><VersionId>null</VersionId><IsLatest>false</IsLatest><LastModified>2020-11-13T10:08:43.765Z</LastModified><ETag>&quot;c662c33d7b4d560f87a167ccca9f5239&quot;</ETag><Size>439</Size><StorageClass>STANDARD</StorageClass><Owner><ID>cl</ID><DisplayName>cl</DisplayName></Owner><Type>Normal</Type></Version></ListVersionsResult>
0
```
And then i specified the keymarker and the latest  version together， i get the latest version only, as 
```
GET /test2/?versions&key-marker=a&version-id-marker=eaVaaCTrIxDnI3ku8FhQ1m.CMPAv7gU&encoding-type=url HTTP/1.1
Host: 10.221.128.15:8080
x-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
Authorization: AWS4-HMAC-SHA256 Credential=cl/20201113//s3/aws4_request, SignedHeaders=amz-sdk-invocation-id;amz-sdk-retry;content-type;host;user-agent;x-amz-content-sha256;x-amz-date, Signature=a20465e5eb410118d54caa2b580eef3cf45510111b3801ecd154f8ed4f5054d5
X-Amz-Date: 20201113T101220Z
User-Agent: aws-sdk-java/1.11.418 Windows_10/10.0 OpenJDK_64-Bit_Server_VM/25.212-b04 java/1.8.0_212-3-redhat
amz-sdk-invocation-id: bf5e2f06-42cc-4bb5-40da-a40a1db2aa39
amz-sdk-retry: 0/0/500
Content-Type: application/octet-stream
Content-Length: 0
Connection: Keep-Alive

HTTP/1.1 200 OK
Transfer-Encoding: chunked
x-amz-request-id: tx000000000000000000017-005fae5bf8-5e98-default
Content-Type: application/xml
Date: Fri, 13 Nov 2020 10:12:08 GMT
Connection: Keep-Alive

<?xml version="1.0" encoding="UTF-8"?><ListVersionsResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><EncodingType>url</EncodingType><Name>test2</Name><Prefix></Prefix><MaxKeys>1000</MaxKeys><IsTruncated>false</IsTruncated><KeyMarker>a</KeyMarker><VersionIdMarker>eaVaaCTrIxDnI3ku8FhQ1m.CMPAv7gU</VersionIdMarker><Version><Key>a</Key><VersionId>null</VersionId><IsLatest>false</IsLatest><LastModified>2020-11-13T10:08:43.765Z</LastModified><ETag>&quot;c662c33d7b4d560f87a167ccca9f5239&quot;</ETag><Size>439</Size><StorageClass>STANDARD</StorageClass><Owner><ID>cl</ID><DisplayName>cl</DisplayName></Owner><Type>Normal</Type></Version></ListVersionsResult>
```
This obviously does not meet the S3 standard.

**Comment by BryceCao** [comment]
@ivancich, Thank you for reminding me, i will add the "Backport" when needed in the future，about the "Pull request ID", how can i add it to tracker after i create the tracker. i tried it before and i failed, is this because i haven't  permission.

**Comment by ivancich** [comment]
So it turns out the ultimate issue is a missing version-marker coming in, which is allowed as shown by Amazon examples, and @cbodley in #37790 has a more elegant solution.

**Pull request closed without merging**